rpc/codec/conn.go

```diff
@@ -9,94 +9,50 @@ import (
 	"encoding/binary"
 	"io"
 	"net"
-	"sync"
-
-	"code.google.com/p/snappy-go/snappy"
 
+	wire "github.com/cockroachdb/cockroach/rpc/codec/wire.pb"
 	"github.com/gogo/protobuf/proto"
 )
 
-type frameBuf struct {
-	size [binary.MaxVarintLen64]byte
+const enableSnappy = true
+
+type decompressFunc func(src []byte, m proto.Message) error
+
+var decompressors = [...]decompressFunc{
+	wire.CompressionType_NONE:   proto.Unmarshal,
+	wire.CompressionType_SNAPPY: snappyDecode,
 }
 
-var framePool = sync.Pool{
-	New: func() interface{} {
-		return &frameBuf{}
-	},
+type baseConn struct {
+	w        *bufio.Writer
+	r        *bufio.Reader
+	c        io.Closer
+	frameBuf [binary.MaxVarintLen64]byte
 }
 
-func sendFrame(w io.Writer, data []byte) (err error) {
+// Close closes the underlying connection.
+func (c *baseConn) Close() error {
+	return c.c.Close()
+}
+
+func (c *baseConn) sendFrame(data []byte) error {
 	// Allocate enough space for the biggest uvarint
-	buf := framePool.Get().(*frameBuf)
-	size := buf.size[:]
+	size := c.frameBuf[:]
 
 	if data == nil || len(data) == 0 {
 		n := binary.PutUvarint(size, uint64(0))
-		err = write(w, size[:n])
-		framePool.Put(buf)
-		return
+		return c.write(c.w, size[:n])
 	}
 
 	// Write the size and data
 	n := binary.PutUvarint(size, uint64(len(data)))
-	err = write(w, size[:n])
-	framePool.Put(buf)
-	if err != nil {
-		return
-	}
-	err = write(w, data)
-	return
-}
-
-func recvProto(r *bufio.Reader, m proto.Message, compressed bool) (err error) {
-	size, err := binary.ReadUvarint(r)
-	if err != nil {
+	if err := c.write(c.w, size[:n]); err != nil {
 		return err
 	}
-	if size != 0 {
-		if r.Buffered() >= int(size) {
-			// Parse proto directly from the buffered data.
-			data, err := r.Peek(int(size))
-			if err != nil {
-				return err
-			}
-			uncompressedData := data
-			if compressed {
-				uncompressedData, err = snappy.Decode(nil, data)
-				if err != nil {
-					return err
-				}
-			}
-			if err := proto.Unmarshal(uncompressedData, m); err != nil {
-				return err
-			}
-			// TODO(pmattis): This is a hack to advance the bufio pointer by
-			// reading into the same slice that bufio.Reader.Peek
-			// returned. In Go 1.5 we'll be able to use
-			// bufio.Reader.Discard.
-			_, err = io.ReadFull(r, data)
-			return err
-		}
-
-		data := make([]byte, size)
-		if _, err := io.ReadFull(r, data); err != nil {
-			return err
-		}
-		if compressed {
-			data, err = snappy.Decode(nil, data)
-			if err != nil {
-				return err
-			}
-		}
-		if err := proto.Unmarshal(data, m); err != nil {
-			return err
-		}
-	}
-	return nil
+	return c.write(c.w, data)
 }
 
-func write(w io.Writer, data []byte) error {
+func (c *baseConn) write(w io.Writer, data []byte) error {
 	for index := 0; index < len(data); {
```

Comment:
True. Also notice that the header and message are written in separate calls to `sendFrame`. But neither of these matter because `net/rpc.{Client,Server}` hold locks while calling `Write{Request,Response}`. See `Client.send` (http://golang.org/src/net/rpc/client.go) and `Server.sendResponse`. I noticed this when looking at the `gob{Client,Server}Codec` implementation and noticing it didn't do any locking. 

I think the `net/rpc` documentation is incorrect. I'll see about sending a change upstream to fix it.